client/mm/mm_arb_market_maker.go

```diff
@@ -0,0 +1,651 @@
+					requiredOnCEX = calc.BaseToQuote(rate, o.lots*mkt.LotSize)
+				} else {
+					requiredOnCEX = o.lots * mkt.LotSize
+				}
+				if requiredOnCEX < remainingCEXBalance {
+					remainingCEXBalance -= requiredOnCEX
+				} else {
+					log.Warnf("rebalance: not enough CEX balance to cover existing order. cancelling.")
+					addCancel(o)
+					remainingCEXBalance = 0
+				}
+			}
+		}
+		if remainingCEXBalance == 0 {
+			log.Debug("rebalance: not enough CEX balance to place new orders")
+			return nil
+		}
+
+		if remainingDEXBalance <= fundingFees {
+			log.Debug("rebalance: not enough DEX balance to pay funding fees")
+			return nil
+		}
+		remainingDEXBalance -= fundingFees
+
+		// For each placement, we check the rate at which the counter trade can
+		// be made on the CEX for the cumulatively required lots * multipliers
+		// of the current and all previous placements. If any orders currently
+		// on the books are outside of the drift tolerance, they will be
+		// cancelled, and if there are less than the required lots on the DEX
+		// books, new orders will be added.
+		placements := make([]*rateLots, 0, len(cfgPlacements))
+		var cumulativeCEXDepth uint64
+		for i, cfgPlacement := range cfgPlacements {
+			cumulativeCEXDepth += uint64(float64(cfgPlacement.Lots*mkt.LotSize) * cfgPlacement.Multiplier)
+			_, extrema, filled, err := a.vwap(sell, cumulativeCEXDepth)
+			if err != nil {
+				log.Errorf("error calculating vwap: %v", err)
+			}
+			if !filled {
+				log.Infof("CEX %s side has < %d %s on the orderbook.", map[bool]string{true: "sell", false: "buy"}[sell], cumulativeCEXDepth, mkt.BaseSymbol)
+				break
+			}
+
+			var placementRate uint64
+			if sell {
+				placementRate = steppedRate(uint64(float64(extrema)*(1+cfg.Profit)), mkt.RateStep)
+			} else {
+				placementRate = steppedRate(uint64(float64(extrema)/(1+cfg.Profit)), mkt.RateStep)
+			}
+
+			ordersForPlacement := existingOrders[i]
+			var existingLots uint64
+			for _, o := range ordersForPlacement {
+				existingLots += o.lots
+				if !withinTolerance(o.rate, placementRate) {
+					addCancel(o)
+				}
+			}
+
+			if cfgPlacement.Lots <= existingLots {
+				continue
+			}
+			lotsToPlace := cfgPlacement.Lots - existingLots
+
+			// TODO: handle redeem/refund fees for account lockers
+			var requiredOnDEX, requiredOnCEX uint64
+			if sell {
+				requiredOnDEX = mkt.LotSize * lotsToPlace
+				requiredOnDEX += sellFees.swap * lotsToPlace
+				requiredOnCEX = calc.BaseToQuote(extrema, mkt.LotSize*lotsToPlace)
+			} else {
+				requiredOnDEX = calc.BaseToQuote(placementRate, lotsToPlace*mkt.LotSize)
+				requiredOnDEX += buyFees.swap * lotsToPlace
+				requiredOnCEX = mkt.LotSize * lotsToPlace
+			}
+			if requiredOnDEX > remainingDEXBalance {
+				log.Debugf("not enough DEX balance to place %d lots", lotsToPlace)
+				continue
+			}
+			if requiredOnCEX > remainingCEXBalance {
+				log.Debugf("not enough CEX balance to place %d lots", lotsToPlace)
+				continue
+			}
```

Comment:
Can also log the balances we have and needed? Is this not a warning? Does the user always to expect to have cfgPlacement.Lots placed?

Reply:
I think that's OK. If the balance has shifted to one side of the market then this is to be expected.